dmoj/executors/filesystem_policies.py

```diff
@@ -0,0 +1,114 @@
+import os
+from typing import List, Union
+
+
+class ACCESS_MODE:
+    NONE = 0
+    EXACT = 1
+    RECURSIVE = 2
+
+
+class Dir:
+    def __init__(self):
+        self.access_mode = ACCESS_MODE.NONE
+        self.map = {}
+
+
+class File:
+    pass
+
+
+class ExactFile:
+    def __init__(self, path: str):
+        self.path = path
+
+
+class ExactDir:
+    access_mode = ACCESS_MODE.EXACT
+
+    def __init__(self, path: str):
+        self.path = path
+
+
+class RecursiveDir:
+    access_mode = ACCESS_MODE.RECURSIVE
+
+    def __init__(self, path: str):
+        self.path = path
+
+
+Rule = Union[ExactFile, ExactDir, RecursiveDir]
+
+
+class FilesystemPolicy:
+    def __init__(self, rules: List[Rule]):
+        self.root = Dir()
+        for rule in rules:
+            self._add_rule(rule)
+
+    def _add_rule(self, rule: Rule):
+        self._assert_rule_type(rule)
+        if rule.path == "/":
+            return self._finalize_root_rule(rule)
+
+        path = rule.path
+        assert os.path.isabs(path), "Rule must specify an absolute path to rule"
+        *directory_path, final_component = path.split("/")[1:]
+
+        node = self.root
+        for component in directory_path:
+            assert component != "", "Must not have empty components in rule to add"
+            new_node = node.map.setdefault(component, Dir())
+            assert isinstance(new_node, Dir), "Cannot descend into non-directory"
+            node = new_node
+
+        self._finalize_rule(node, final_component, rule)
+
+    def _assert_rule_type(self, rule: Rule):
+        if os.path.exists(rule.path):
+            is_dir = os.path.isdir(rule.path)
+            if isinstance(rule, ExactFile):
+                assert not is_dir, f"Can't apply file rule to directory {rule.path}"
+            else:
+                assert is_dir, f"Can't apply directory rule to non-directory {rule.path}"
+
+    def _finalize_root_rule(self, rule: Rule):
+        assert not isinstance(rule, ExactFile), "Root is not a file"
+        self._finalize_directory_rule(self.root, rule)
+
+    def _finalize_rule(self, node: Dir, final_component: str, rule: Rule):
+        assert final_component != "", "Must not have trailing slashes in rule path"
+        if isinstance(rule, ExactFile):
+            new_node = node.map.setdefault(final_component, File())
+            assert isinstance(new_node, File), "Can't add ExactFile: Dir rule exists"
+        else:
+            new_node = node.map.setdefault(final_component, Dir())
+            assert isinstance(new_node, Dir), "Can't add rule: File rule exists"
+            self._finalize_directory_rule(new_node, rule)
+
+    def _finalize_directory_rule(self, node: Dir, rule: Union[ExactDir, RecursiveDir]):
+        node.access_mode = max(node.access_mode, rule.access_mode)  # Allow the more permissive rule
+
+    # `path` should be a canonical path. The output of `realpath` is appropriate
```

Comment:
I think we ought to move the path normalization into `check`, since that avoids the class of bugs where we pass an un-normalized path to `check`. Until now it had to be elsewhere, since we were using regex directly and didn't have a wrapper class.

(This would be a good candidate for a unit test too.)

Reply:
Or maybe not, it's a bit annoying that we need the cwd for this.

Reply:
I'm not opposed to moving the logic in here, the CWD fetching is relatively straightforward, right? `check` could take `dir_fd` and `file_path`

Reply:
If `dir_fd` is `AT_FDCWD` you need to read the `cwd` of the calling process (https://github.com/DMOJ/judge-server/blob/e20c1d61b3a103fc80ddc5a12122b6b262a2c0a4/dmoj/cptbox/isolate.py#L268), so you need access to debugger instance, and have separate logic for Linux/FreeBSD to get the cwd.

Reply:
Yeah that's definitely annoying. We could pass the directory path and file path and have the filesystem policy handle joining

Reply:
This will be resolved after #882 is merged.